### DISCUSSION TITLE
Nuxt 3 with Fabricjs results in crashed serverless function due to missing zlib

### DISCUSSION THREAD

**Discussion by MatthewLeeCode** [vercel/community#4759] [Help] [answered]
### Summary

Hello,

First time using Vercel. I'm attempting to build my Nuxt 3 app which includes fabricjs as a dependency. Unfortunately I have received the error 

`500: INTERNAL_SERVER_ERROR
Code: FUNCTION_INVOCATION_FAILED
ID: syd1::rtg4q-1699334988691-d2d2302e593`

Looking at the logs it appears to be because zlib isn't installed

```
Error: /lib64/libz.so.1: version `ZLIB_1.2.9' not found (required by /var/task/node_modules/canvas/build/Release/libpng16.so.16)
    at Module._extensions..node (node:internal/modules/cjs/loader:1340:18)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
    at t.<computed>.e._load (/var/task/___vc/__launcher/__launcher.js:14:2299)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (/var/task/node_modules/canvas/lib/bindings.js:3:18)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32) {
  code: 'ERR_DLOPEN_FAILED'
}
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: /lib64/libz.so.1: version `ZLIB_1.2.9' not found (required by /var/task/node_modules/canvas/build/Release/libpng16.so.16)","reason":{"errorType":"Error","errorMessage":"/lib64/libz.so.1: version `ZLIB_1.2.9' not found (required by /var/task/node_modules/canvas/build/Release/libpng16.so.16)","code":"ERR_DLOPEN_FAILED","stack":["Error: /lib64/libz.so.1: version `ZLIB_1.2.9' not found (required by /var/task/node_modules/canvas/build/Release/libpng16.so.16)","    at Module._extensions..node (node:internal/modules/cjs/loader:1340:18)","    at Module.load (node:internal/modules/cjs/loader:1119:32)","    at Module._load (node:internal/modules/cjs/loader:960:12)","    at t.<computed>.e._load (/var/task/___vc/__launcher/__launcher.js:14:2299)","    at Module.require (node:internal/modules/cjs/loader:1143:19)","    at require (node:internal/modules/cjs/helpers:119:18)","    at Object.<anonymous> (/var/task/node_modules/canvas/lib/bindings.js:3:18)","    at Module._compile (node:internal/modules/cjs/loader:1256:14)","    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)","    at Module.load (node:internal/modules/cjs/loader:1119:32)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: /lib64/libz.so.1: version `ZLIB_1.2.9' not found (required by /var/task/node_modules/canvas/build/Release/libpng16.so.16)","    at process.<anonymous> (file:///var/runtime/index.mjs:1276:17)","    at process.emit (node:events:529:35)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```

which appears to be a fabricjs requirement.

Unsure what to do from here.

Thanks

### Example

https://gamecoach.vercel.app/

### Steps to Reproduce

Simply went through the Nuxt3 deploy wizard. Nothing fancy. Fabricjs is a required npm dependency

**Comment by amyegan** [comment]
It looks like the build container does not have the needed package for your project. You can add it by overriding the Install Command in [Build and Development Settings](https://vercel.com/docs/deployments/configure-a-build#build-and-development-settings).

You can use `yum` in the Install Command to install other native dependencies to the build image. A command like `yum install [package-name] -y && npm install` would install the package, default 'yes' to any installation questions, and follow up with the usual `npm install`. 

This will override the standard install command (npm install or yarn install) and ensure that when the npm dependencies are built, the required native dependencies are present.

**Comment by marctaule** [comment]
@MatthewLeeCode did you find a solution? I tried a lot of things I found on Google but nothing works :(

**Comment by MatthewLeeCode** [comment]
Hi @marctaule, wish I had better news but unfortunately I couldn't get it working. There appears to be an issue with Fabricjs node dependencies and the Vercel environment. The closest I got to it working was installing 'https://github.com/finom/fabric-browser' but then I received a different problem 'No DOMParser'. 

Thanks for your help @amyegan it did give me some more options to try but unfortunately installing zlib through that method didn't work.

My solution was to go with Firebase Hosting which worked. I'd be willing to give this another go because I like the Vercel integration with Supabase so if anyone has other ideas let me know!

**Comment by spaceassault** [comment, marked as answer]
@MatthewLeeCode I had the same issue. I am using Remix and was able to solve it by using a client-only component (https://remix.run/docs/en/main/file-conventions/-client). I don't know much about Nuxt 3 but you might try this approach in Nuxt 3 with client-only components. https://nuxt.com/docs/api/components/client-only

**Comment by MatthewLeeCode** [comment]
@spaceassault thanks for your comment, It looks like that worked! (And by worked I mean I hit different, unrelated problems. Progress!)

I just wrapped the canvas component with 'Client Only'. I suppose SSR was attempting to load the canvas with fabricjs on the server side and caused the issues above.

```
<template>
    <div>
        <ClientOnly>
            <canvas-board :board_link="board.link" ></canvas-board>
        </ClientOnly>
    </div>
</template>
```

**Reply by marctaule** [reply]
Hi @MatthewLeeCode! 
Sorry for replying so late
I created a plugin on the client with Fabric.js and it worked :)

`import { fabric } from 'fabric'
export default defineNuxtPlugin(() => {
  window.fabric = window.fabric = fabric
})
`

Then you can use fabric inside your page.

**Reply by matifanger** [reply]
I tried it but now this error appears:

`'fabric' refers to a UMD global, but the current file is a module. Consider adding an import instead.ts(2686)`
